### PR TITLE
generate oneof types

### PR DESCRIPTION
**Pull request by pongad**
For each oneof group in a message,
we create a proxy type that sets one field in the group.
These types will be later be used by toolkit to flatten
oneof fields.

Testing with pubsub, the generated code compiles and looks
good enough for early feedback.
However, we should test with other APIs, and there are
some finer points I'd like to discuss more, hence WIP.

The oneof types are nested in the message type,
because the oneof group is scoped in the message in protobuf.

To reviewers: new code starts at line 1088 of the baseline.

### CODE REVIEW

**Comment by michaelbausor** [comment]
This looks really good!

I am trying to think, is there ever any case for someone to need to create their own subclass of the oneof type? I think not - one possibility is when a new field is added to the oneof. But unlike the resource name case, with a oneof we will automatically generate support for it as soon as we regenerate the proto, with no gapic config update required.

**Comment by garrettjonesgoogle** [comment]
@michaelbausor that would be like letting people subclass protos because new fields are added, but protos are currently final. I don't think we should make a special case for oneofs.

File: gapic_plugin.py [hunk 1]
```diff
@@ -99,6 +101,20 @@ def generate_get_set_injection(response, gapic_config, request, java_package):
...
+      for i, oneof in enumerate(item.oneof_decl):
+        f = response.file.add()
+        f.name = filename
+        f.insertion_point = 'builder_scope:' + package + '.' + item.name
+        f.content = renderer.render(oneof_fields.InsertBuilderOneof(item, oneof.name, oneofs[i], type_map))
+
+        f = response.file.add()
+        f.name = filename
+        f.insertion_point = 'class_scope:' + package + '.' + item.name
+        f.content = renderer.render(oneof_fields.InsertClassOneof(item, oneof.name, oneofs[i], type_map))
```
Comment (garrettjonesgoogle) [inline gapic_plugin.py]: This method is getting a little too long, and there is an "indentation cliff" in the middle of it that makes it hard to match up the blocks.

**Comment by pongad** [comment]
@garrettjonesgoogle Does this look any better to you? We can split the func but I think that makes things even more confusing.

**Comment by michaelbausor** [comment]
These change LGTM pending finalizing the design

**Comment by pongad** [comment]
@garrettjonesgoogle @michaelbausor PTAL. I think this should resolve all blocking design problems. Non blocking stuff like making it work with resource name types can be done separately.

**Comment by michaelbausor** [comment]
LGTM

**Comment by pongad** [comment]
This plugin should soon be retired. Closing this.

**Pull request closed without merging**